lib/chef/provider/package/apt.rb

```diff
@@ -37,23 +38,22 @@ def initialize(new_resource, run_context)
         def load_current_resource
           @current_resource = Chef::Resource::AptPackage.new(new_resource.name)
           current_resource.package_name(new_resource.package_name)
-          current_resource.version(get_current_versions)
-          current_resource
-        end
 
-        def define_resource_requirements
-          super
+          if source_files_exist?
+            @candidate_version = get_candidate_versions
```

Comment:
I'm skeptical the right thing is going to happen here, and it doesn't appear to be in testing.

Because there are no changes in this code-path, we're getting the candidate versions from `apt-cache policy` which doesn't look at the local deb package at all. It's going to, at best, determine the candidate version from what would come from the apt sources, however this appears to be broken in this branch:

```
$ cat apache.rb
package 'apache2_2.4.41-4ubuntu3.8_amd64.deb'
```

```
$ sudo chef-apply apache.rb
Recipe: (chef-apply cookbook)::(chef-apply recipe)
  * apt_package[apache2_2.4.41-4ubuntu3.8_amd64.deb] action install
    * No candidate version available for apache2_2.4.41-4ubuntu3.8_amd64.deb
    ================================================================================
    Error executing action `install` on resource 'apt_package[apache2_2.4.41-4ubuntu3.8_amd64.deb]'
    ================================================================================

    Chef::Exceptions::Package
    -------------------------
    No candidate version available for apache2_2.4.41-4ubuntu3.8_amd64.deb

    Resource Declaration:
    ---------------------
    # In apache.rb

      1: package 'apache2_2.4.41-4ubuntu3.8_amd64.deb'

    Compiled Resource:
    ------------------
    # Declared in apache.rb:1:in `run_chef_recipe'

    apt_package("apache2_2.4.41-4ubuntu3.8_amd64.deb") do
      package_name "apache2_2.4.41-4ubuntu3.8_amd64.deb"
      action [:install]
      default_guard_interpreter :default
      declared_type :package
      cookbook_name "(chef-apply cookbook)"
      recipe_name "(chef-apply recipe)"
    end

    System Info:
    ------------
    chef_version=17.4.20
    platform=ubuntu
    platform_version=20.04
    ruby=ruby 2.7.0p0 (2019-12-25 revision 647ee6f091) [x86_64-linux-gnu]
    program_name=/usr/local/bin/chef-apply
    executable=/usr/local/bin/chef-apply

[2021-11-23T15:38:34-05:00] FATAL: Stacktrace dumped to /var/chef/cache/chef-stacktrace.out
[2021-11-23T15:38:34-05:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
[2021-11-23T15:38:34-05:00] FATAL: Chef::Exceptions::Package: apt_package[apache2_2.4.41-4ubuntu3.8_amd64.deb] ((chef-apply cookbook)::(chef-apply recipe) line 1) had an error: Chef::Exceptions::Package: No candidate version available for apache2_2.4.41-4ubuntu3.8_amd64.deb
```

Reply:
I was wondering why with all the methods that were copied from `Package::Dpkg`, #name_candidate_version wasn't  copied but that makes sense now given that this whole section of code would need to be refactored.